### PR TITLE
(Depends on PR #18 #17) RNN Decoder with attention

### PR DESCRIPTION
**Pull request by aasseman**
(Depends on PR #18  #17)

Adding RNN Decoder w/ attention, as per [this pytorch tutorial](https://pytorch.org/tutorials/intermediate/seq2seq_translation_tutorial.html). Itself based on the paper [*Neural Machine Translation by Jointly Learning to Align and Translate*](https://arxiv.org/abs/1409.0473).

### CODE REVIEW

**Comment by aasseman** [comment]
Seems to work well enough for now to run `wikitext_language_modeling_encoder_attndecoder.yml`, which is a basic copy task.
Supports only single layer GRU for now. Will probably not try extending that. Maybe multiple are useful?

Will try to add the translation task from the pytorch tutorial, as to compare apples to apples if my implementation is doing what it should.

**Comment by tkornut** [comment]
This pull request **introduces 1 alert** when merging df60fb46454347746e1ae192b9e8f52a838870d8 into aed980d5a48b2cb98dc460e8e58ddecced08813e - [view on LGTM.com](https://lgtm.com/projects/g/IBM/pytorchpipe/rev/pr-e309e8aba7b609d9bc93546883f18d31edb7779a)

**new alerts:**

* 1 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by tkornut** [comment]
This pull request **introduces 1 alert** when merging 0138ef22fe549f6fece86a865d1c2940846b9c76 into aed980d5a48b2cb98dc460e8e58ddecced08813e - [view on LGTM.com](https://lgtm.com/projects/g/IBM/pytorchpipe/rev/pr-f55cca4c74902161473971f9fd3684ffaf84602e)

**new alerts:**

* 1 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by aasseman** [comment]
Ready for review.
Got good results on the translation task.
One thing that could be checked (visually), if time permits, is the attention matrix of the GRU decoder, to really make sure that we get similar results to the paper.

File: configs/wikitext/wikitext_language_modeling_seq2seq.yml [hunk 1]
```diff
@@ -94,14 +96,17 @@ pipeline:
...
-    type: RecurrentNeuralNetwork
+    type: Attn_Decoder_RNN
     priority: 4
-    initial_state: Input
-    hidden_size: 300
-    num_layers: 3
+    hidden_size: 50
+    num_layers: 1
     use_logsoftmax: False
+<<<<<<< Updated upstream
```
Comment (tkornut) [inline configs/wikitext/wikitext_language_modeling_seq2seq.yml]: Hmmmm, I guess the "updated upstream" is the right one... but not sure..

Comment (aasseman) [reply configs/wikitext/wikitext_language_modeling_seq2seq.yml]: I guess I got mixed up with my stashes. I checked-out the `develop` version of the two files in question, and that solved it.

**Pull request merged**